### DISCUSSION TITLE
Support additional build arguments (not --build-arg)

### DISCUSSION THREAD

**Discussion by Roemer** [devcontainers/spec#301] [Ideas]
Currently, `build` has an object `args` which can be used to pass `--build-arg` to the `docker build` command.
But it is often needed, to also pass other command line arguments, like `--network` or `--add-host`. This is currently not possible in an easy way.

I see two possibilities:
1. Rename `args` to `vars` or `variables` as they just set build-time variables and re-use `args` for the additional cli-arguments. This would be more in-line with `runArgs` but introduces a heavy breaking change.
2. Add a new property `cliArgs` or `buildArgs` (would then be `build.buildArgs`) to the `build` section which are then passed to `docker build`. This would just be an extension and not break any existing dev-container definitions.

This would solve the issue https://github.com/microsoft/vscode-remote-release/issues/3545 which already has almost 100 upvotes.

**Comment by sliekens** [comment]
I would also like to mention it's not currently possible to pass ANY build args when using a combination  of `image` and `features`, even though this will for sure do a `docker build` .

``` json
{
  "image": "mcr.microsoft.com/devcontainers/base",
  "features": {
    "ghcr.io/devcontainers/features/node:1": "lts"
  },
  "build": { // ERROR: property build is not allowed
    "cacheFrom": "ghcr.io/owner/repo/devcontainer:latest"
  }
}
```

As you can see, property `build` is not allowed when using `image` although `features` would benefit from caching from a pre-built image to speed up builds.

**Comment by scalpel4k** [comment]
we have the "runArgs" for 'docker run', why can't the same be done for 'docker build' using let's say "buildArgs" ?

**Reply by Roemer** [reply]
`BuildArgs` have a special meaning in `docker build`. They act like placeholders and are used in the  `Dockerfile` with `ARG`. This is what we already have with `build.args`.
What we want here is actually cli arguments passed to the docker command. So we should call them `build.buildArgs` (actually a bit misleading) or better `build.cliArgs` or `build.commandArgs` or something like that.

**Reply by hakontonne** [reply]
Why not call it `build.options`? That would keep this in line with the official docker documentation that explicitly names them options:
`docker build [OPTIONS] PATH | URL | -`

https://docs.docker.com/engine/reference/commandline/build/#options

**Reply by Roemer** [reply]
Yep, that's what my proposal, ehm, proposes: options (https://github.com/devcontainers/spec/pull/328)

**Reply by chrmarti** [reply]
This is in progress now: https://github.com/devcontainers/spec/pull/324 https://github.com/devcontainers/cli/issues/85. Thanks @Roemer !
